### ISSUE TITLE
Generate `.ts` files instead of `.d.ts` files

### ISSUE THREAD

**Issue by geigerzaehler** [ethereum-ts/TypeChain#430]
Instead of generating declaration files with the `.d.ts` extension TypeChain should generate `.ts` files for contract interfaces.

TypeScript [does not copy declaration files](https://stackoverflow.com/questions/56018167/typescript-does-not-copy-d-ts-files-to-build) when building a project. This leads to issues when compiling contract bindings in a package to distribute them: Dependents cannot resolve the declaration files because they are not in the correct location.

It should be easy to generate `.ts` files instead of `.d.ts` files since contract interfaces only include type information and no executable code. I would probably also be a good idea to use [`import type`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) in the generated files.

**Comment by krzkaczor** [comment]
Yes, the fact that tsc doesn't copy declaration files can be a PITA but I am not sure if changing `.d.ts` to `.ts` is a simple change. Initial idea for using `.d.ts` was to ensure that typechain doesn't leave a runtime trace (just provides types for existing runtime objects). I am not sure if this is still the case as now we have a mix of ts and .d.ts files... Perhaps this should be changes in next major version. 

@quezak @zemse thoughts on this?

**Comment by quezak** [comment]
I have no objections against this change.

Keeping no runtime trace makes sense for the frontend where the bundle size is important, but then TypeChain's types are invisible at runtime. Though I'm only using it on the backend, IMO the advantages of runtime types are also important, for creating helper functions, easy instantiation of the correct contract sub-class, logging, etc. I remember I had to add some more copypaste and boilerplate in my projects because that's not possible with just the declaration files.

**Comment by zemse** [comment]
Yeah, I think the change suggested is good. (Until now I used to [manually copy](https://github.com/KMPARDS/eraswap-sdk/blob/master/package.json#L11) the `.d.ts` files to build dir). From typescript developers (https://github.com/microsoft/TypeScript/issues/40517#issuecomment-692216717), I understand that purpose of `.d.ts` is used to introduce typings to existing js files/packages without typings. That is the reason they have given for not copying `.d.ts` files from `src` to build output. So in our case, I feel `.ts` files are more suitable.

If we do the type imports even if it's `.ts`, it should not leave any run time trace?

```ts
import type {
  DataTypesInput,
  DataTypesInputInterface,
} from "../DataTypesInput";
```

I just tried in one of my old react projects having typechain output in src, with `.t.ds` build size is `3,004,261 bytes`. When I change it to `.ts` the build is `2,998,113 bytes`. Weirdly the build size decreases slightly (I retried to see same result). Anyway, it seems there is no hit on build size when we have `.ts` + type imports to me.

**Comment by dave4506** [comment]
Is there any follow up re this? We been running into build issues with `.t.ds` files.

**Comment by krzkaczor** [comment]
It's planned to be fixed in the next major version.

**Comment by silasdavis** [comment]
@krzkaczor would you like some help on this? Do you foresee any particular difficulties?

**Comment by krzkaczor** [comment]
@silasdavis next month at the latest it's gonna be done. Sorry for the delay.

**Comment by silasdavis** [comment]
No need to apologise you owe me nada!

Here is my hack that works for me on ethers-v5: https://github.com/dethcrypto/TypeChain/pull/486/files

I could try and expand that a bit but sounds like you have it in hand.

**Comment by freeatnet** [comment]
Looks like this change implicitly enabled `tsc` lint checks for the generated type files — including `noUnusedLocals` (TS6133), which is triggered when on unused imports. It seems like there's some logic to avoid unused imports [in contract factory code generation](https://github.com/dethcrypto/TypeChain/blob/ba4c18ad49836f56f65dfb8e662786bd40e0a426/packages/target-ethers-v5/src/codegen/index.ts#L145). Would it make sense to do the same for type files as well?
